util/src/main/java/org/killbill/billing/util/security/shiro/realm/KillBillJndiLdapRealm.java

```diff
@@ -190,27 +193,28 @@ private Set<String> findLDAPGroupsForUser(final String userName, final LdapConte
         return new HashSet<>(finalGroupsNames);
     }
 
-    private Collection<String> extractGroupNamesFromSearchResult(final SearchResult searchResult) {
-        // Get all attributes for that group
+    @VisibleForTesting
+    Collection<String> extractGroupNamesFromSearchResult(final SearchResult searchResult) {
+        // Extract the group name from the attribute
         final Iterator<? extends Attribute> attributesIterator = searchResult.getAttributes().getAll().asIterator();
-
-        return Iterators.toStream(attributesIterator)
-                        // Find the attribute representing the group name
-                        .filter(attribute -> groupNameId.equalsIgnoreCase(attribute.getID()))
-                        // Extract the group name from the attribute
-                        // Note: at this point, groupNameAttributesIterator should really contain a single element
-                        .map(groupNameAttribute -> {
-                            try {
-                                final NamingEnumeration<?> enumeration = groupNameAttribute.getAll();
-                                return enumeration.asIterator();
-                            } catch (final NamingException namingException) {
-                                log.warn("Unable to read group name(s)", namingException);
-                                return null;
-                            }
-                        })
-                        .filter(Objects::nonNull)
-                        .map(Object::toString)
-                        .collect(Collectors.toUnmodifiableSet());
+        final Collection<String> groups = Iterators.toStream(attributesIterator)
+                                                   // Find the attribute representing the group name
+                                                   .filter(attribute -> groupNameId.equalsIgnoreCase(attribute.getID()))
+                                                   .flatMap(groupNameAttribute -> {
+                                                 try {
+                                                     final NamingEnumeration<?> enumeration = groupNameAttribute.getAll();
+                                                     final List<?> enumerationAsList = Collections.list(enumeration);
+                                                     final List<String> values = new ArrayList<>();
+                                                     enumerationAsList.forEach(e -> values.addAll((Collection) e));
```

Comment:
I'm concerned about the cast here: we don't really know what `enumeration` is (could be LDAP implementation specific?).

How about?
```
return enumerationAsList.stream().flatMap(value -> {
            if (value instanceof Collection) {
                return ((Collection<?>) value).stream().map(Object::toString);
            } else {
                return Stream.of(value.toString());
            }
        });
```